### PR TITLE
Rollup update buckets

### PR DESCRIPTION
**Pull request by Mikelle**
Close https://github.com/hermeznetwork/hermez-node/issues/643

### CODE REVIEW

File: common/ethrollup.go [hunk 1]
```diff
@@ -132,10 +132,12 @@ func (c *RollupConstants) FindVerifierIdx(MaxTx, NLevels int64) (int, error) {
 // BucketParams are the parameter variables of each Bucket of Rollup Smart
```
[the remark this answers is not in the document]
Comment (ed255) [reply common/ethrollup.go:132]: I've seen that `common.RollupConstNumBuckets` is still used in several fixed-length arrays.  All those arrays should be converted to slices, and managed as variable length (with proper allocations and so on).

**Review by ed255 — CHANGES_REQUESTED** [review]

File: eth/rollup.go [hunk 2]
```diff
@@ -493,8 +496,8 @@ func (c *RollupClient) RollupL1UserTxERC20ETH(fromBJJ babyjub.PublicKeyComp, fro
 				auth.Value = depositAmount
 			}
 			var permit []byte
-			return c.hermez.AddL1Transaction(auth, babyPubKey, fromIdxBig, uint16(depositAmountF),
-				uint16(amountF), tokenID, toIdxBig, permit)
+			return c.hermez.AddL1Transaction(auth, babyPubKey, fromIdxBig, big.NewInt(0).SetUint64(uint64(depositAmountF)),
```
Comment (Pantani) [inline eth/rollup.go:499]:
```suggestion
			return c.hermez.AddL1Transaction(auth, babyPubKey, fromIdxBig, big.NewInt(int64(depositAmountF)),
```

File: eth/rollup.go [hunk 3]
```diff
@@ -493,8 +496,8 @@ func (c *RollupClient) RollupL1UserTxERC20ETH(fromBJJ babyjub.PublicKeyComp, fro
...
+				big.NewInt(0).SetUint64(uint64(amountF)), tokenID, toIdxBig, permit)
```
Comment (Pantani) [inline eth/rollup.go:500]:
```suggestion
				big.NewInt(int64(amountF)), tokenID, toIdxBig, permit)
```

File: eth/rollup.go [hunk 4]
```diff
@@ -542,7 +545,7 @@ func (c *RollupClient) RollupL1UserTxERC20Permit(fromBJJ babyjub.PublicKeyComp,
 			signature, _ := c.client.ks.SignHash(*c.client.account, digest)
 			permit := createPermit(owner, spender, amount, deadline, digest, signature)
 			return c.hermez.AddL1Transaction(auth, babyPubKey, fromIdxBig,
-				uint16(depositAmountF), uint16(amountF), tokenID, toIdxBig, permit)
+				big.NewInt(0).SetUint64(uint64(depositAmountF)), big.NewInt(0).SetUint64(uint64(amountF)), tokenID, toIdxBig, permit)
```
Comment (Pantani) [inline eth/rollup.go:548]:
```suggestion
				big.NewInt(int64(depositAmountF)), big.NewInt(int64(amountF)), tokenID, toIdxBig, permit)
```

File: common/ethrollup.go [hunk 5]
```diff
@@ -132,10 +132,12 @@ func (c *RollupConstants) FindVerifierIdx(MaxTx, NLevels int64) (int, error) {
...
-	CeilUSD             *big.Int
-	Withdrawals         *big.Int
-	BlockWithdrawalRate *big.Int
-	MaxWithdrawals      *big.Int
+	CeilUSD         *big.Int
+	BlockStamp      *big.Int
+	Withdrawals     *big.Int
+	RateBlocks      *big.Int
+	RateWithdrawals *big.Int
+	MaxWithdrawals  *big.Int
```
[the remark this answers is not in the document]
Comment (Mikelle) [reply common/ethrollup.go:138]: Decided to stick with *big.Int, bcs Hermez.go also saves vars in this format

File: api/swagger.yml [hunk 6]
```diff
@@ -2731,23 +2731,33 @@ components:
                 type: string
                 description: Max USD value
                 example: "1000"
+              blockStamp:
+                type: string
+                description: Block stamp
```
[the remark this answers is not in the document]
Comment (Mikelle) [reply api/swagger.yml]: Fixed

File: eth/rollup.go [hunk 7]
```diff
@@ -146,14 +147,16 @@ type RollupEventUpdateWithdrawalDelay struct {
...
+	BlockStamp      *big.Int
+	Withdrawals     *big.Int
+	RateBlocks      *big.Int
+	RateWithdrawals *big.Int
+	MaxWithdrawals  *big.Int
 }
 
 type rollupEventUpdateBucketsParametersAux struct {
-	ArrayBuckets [common.RollupConstNumBuckets][4]*big.Int
+	ArrayBuckets []*big.Int
```
[the remark this answers is not in the document]
Comment (Mikelle) [reply eth/rollup.go:159]: Changed to []*big.Int, bcs event returns data in this format

File: eth/rollup.go [hunk 8]
```diff
@@ -907,10 +910,16 @@ func (c *RollupClient) RollupEventsByBlock(blockNum int64,
 				return nil, tracerr.Wrap(err)
 			}
 			for i, bucket := range bucketsParametersAux.ArrayBuckets {
-				bucketsParameters.ArrayBuckets[i].CeilUSD = bucket[0]
-				bucketsParameters.ArrayBuckets[i].Withdrawals = bucket[1]
-				bucketsParameters.ArrayBuckets[i].BlockWithdrawalRate = bucket[2]
-				bucketsParameters.ArrayBuckets[i].MaxWithdrawals = bucket[3]
+				bucket, err := c.hermez.UnpackBucket(c.opts, bucket)
```
[the remark this answers is not in the document]
Comment (Mikelle) [reply eth/rollup.go:915]: fixed

File: eth/rollup.go [hunk 9]
```diff
@@ -60,7 +60,7 @@ type RollupEventInitialize struct {
 
 // RollupVariables returns the RollupVariables from the initialize event
 func (ei *RollupEventInitialize) RollupVariables() *common.RollupVariables {
-	var buckets [common.RollupConstNumBuckets]common.BucketParams
+	buckets := make([]common.BucketParams, 1)
```
Comment (ed255) [inline eth/rollup.go:63]: In the initialization of `InstantWithdrawManager.sol` (where the buckets are defined), the buckets are not touched.  This means that on initalization, there are 0 buckets initially, so buckets should be an empty slice. Example
```go
buckets := []common.BucketsParams{}
```

https://github.com/hermeznetwork/contracts/blob/d439b23a81e64f6cd65869022d06d7ef85729632/contracts/hermez/Hermez.sol#L223
https://github.com/hermeznetwork/contracts/blob/d439b23a81e64f6cd65869022d06d7ef85729632/contracts/hermez/lib/InstantWithdrawManager.sol#L58

**Review by ed255 — CHANGES_REQUESTED** [review]

File: eth/rollup.go (hunk 9 above)
Comment (Mikelle) [reply eth/rollup.go]: Fixed

**Comment by ed255** [comment]
I have two requests:
1. Please update `eth/README.md` with the branch name and commit hash that you used with the updated contracts
2. I have run the eth integration tests and there's a failure:
```
--- FAIL: TestRollupWithdrawMerkleProof (0.25s)
panic: runtime error: index out of range [0] with length 0 [recovered]
        panic: runtime error: index out of range [0] with length 0

goroutine 425 [running]:
testing.tRunner.func1.2(0xbfd960, 0xc0004275f0)
        /usr/lib/go/src/testing/testing.go:1144 +0x332
testing.tRunner.func1(0xc0006ce900)
        /usr/lib/go/src/testing/testing.go:1147 +0x4b6
panic(0xbfd960, 0xc0004275f0)
        /usr/lib/go/src/runtime/panic.go:965 +0x1b9
github.com/hermeznetwork/hermez-node/eth.TestRollupWithdrawMerkleProof(0xc0006ce900)
        /home/dev/git/iden3/hermez/hermez-node/eth/rollup_test.go:1028 +0x7b6
testing.tRunner(0xc0006ce900, 0xde2290)
        /usr/lib/go/src/testing/testing.go:1194 +0xef
created by testing.(*T).Run
        /usr/lib/go/src/testing/testing.go:1239 +0x2b3
exit status 2
FAIL    github.com/hermeznetwork/hermez-node/eth        9.510s
```
Have you encountered this error?

It seems that failure happens because of the following reason: the `updateBucketWithdraw` event is not emitted during the test.  During the last month there was an update in the smart contracts that made the `InstantWithdrawManager.sol` not always emit the `updateBucketWithdraw` event: https://github.com/hermeznetwork/contracts/blob/d439b23a81e64f6cd65869022d06d7ef85729632/contracts/hermez/lib/InstantWithdrawManager.sol#L81
There are several reasons for `_processInstantWithdrawal` to not emit events:
a. amountUSD = 0
b. appropiate bucket for amountUSD not found
c. withdrawals = false

I think we are seeing case b), and to fix that I suggest changing the bucket parameters so that there's a bucket that accommodates the amount in USD in the withdrawal after applying the exchange.  Another option is to modify the token exchange so that the amount in USD fits a current bucket.

**Review by ed255 — DISMISSED** [review]

**Review by ed255 — APPROVED** [review]

**Review by arnaucube — COMMENTED** [review]
I understand that this PR updates the smart contract interfaces & db to the last version of the smart contracts code (https://github.com/hermeznetwork/hermez-node/issues/643), but there are 13 commits with small changes that combining those 12 commits adds the functionality.
Could you check the guidelines (https://www.notion.so/iden3/Github-usage-agreements-68345124b60d445bac7de4ffff48eb40#98382d107ac44f32ab75cf10e10be5aa) and try to group the commits? In order to have better traceability of the git history.

**Comment by Mikelle** [comment]
> I understand that this PR updates the smart contract interfaces & db to the last version of the smart contracts code (#643), but there are 13 commits with small changes that combining those 13 commits adds the functionality.
> Could you check the guidelines (https://www.notion.so/iden3/Github-usage-agreements-68345124b60d445bac7de4ffff48eb40#98382d107ac44f32ab75cf10e10be5aa) and try to group the commits? In order to have better traceability of the git history.

Could you use squash and merge, when you will merge the PR?

File: api/api_test.go [hunk 10]
```diff
@@ -499,7 +499,7 @@ func TestMain(m *testing.M) {
 	// Slot 7
 	nextForgers[6].Coordinator = nonBootForger
 
-	var buckets [common.RollupConstNumBuckets]common.BucketParams
+	buckets := make([]common.BucketParams, 5)
```
Comment (arnaucube) [inline api/api_test.go:502]: Instead of `5`, could be used the defined constant `common.RollupConstNumBuckets` (which has value `5` also)?

**Pull request merged**